lib/ansible/inventory/__init__.py

```diff
@@ -47,165 +44,483 @@
+        display.debug('set %s for %s' % (varname, entity))
+
+
+    def add_child(self, group, child):
+        ''' Add host or group to group '''
+
+        if group in self.groups:
+            g = self.groups[group]
+            if child in self.groups:
+                g.add_child_group(self.groups[child])
+            elif child in self.hosts:
+                g.add_host(self.hosts[child])
+            else:
+                raise AnsibleError("%s is not a known host nor group" % child)
+            self._groups_dict_cache = {}
+            display.debug('Group %s now contains %s' % (group, child))
+        else:
+            raise AnsibleError("%s is not a known group" % group)
+
+    def get_groups_dict(self):
+        """
+        We merge a 'magic' var 'groups' with group name keys and hostname list values into every host variable set. Cache for speed.
+        """
+        if not self._groups_dict_cache:
+            for (group_name, group) in iteritems(self.groups):
+                self._groups_dict_cache[group_name] = [h.name for h in group.get_hosts()]
+
+        return self._groups_dict_cache
+
+
+class InventoryManager(object):
+    ''' Creates and manages inventory '''
+
+    def __init__(self, loader, sources=None):
+
+        # base objects
+        self._loader = loader
+        self._inventory = Inventory()
+
+
+        # a list of host(names) to contain current inquiries to
+        self._restriction = None
+        self._subset = None
+
+        # caches
+        self._pattern_cache = {}        # resolved host patterns
+        self._inventory_generators = {}  # for retrieving inventory
+        self._inventory_processors = {} # for modifyng inventory
+
+        # the inventory dirs, files, script paths or lists of hosts
+        if sources is None:
+            self._sources = [ C.DEFAULT_HOST_LIST ]
+        else:
+            if isinstance(sources, string_types):
+                self._sources = [ sources ]
+            else:
+                self._sources = sources
+
+        # get to work!
+        self.parse_sources()
+
+    def __getattr__(self, attr):
+        ''' allow using inventory data methods directly '''
+        display.debug('inventory indirection used for %s' % attr)
+        return getattr(self._inventory, attr)
+
+    def _parse_host_list(self, host_list):
+        ''' parses cli supplied host lists into inventory'''
+
+        for h in host_list:
+            if h:
+                try:
+                    (host, port) = parse_address(h, allow_ranges=False)
+                except AnsibleError as e:
+                    display.vvv("Unable to parse address from hostname, leaving unchanged: %s" % to_text(e))
+                    host = h
+                    port = None
+
+                if host not in self._inventory.hosts:
+                    self._inventory.add_host(host, group='ungrouped', port=port)
+            else:
+                display.warning("Skipping invalid hostname: %s" % to_text(h))
```

Comment:
`h` could be an empty string, therefore `%r` could be used here instead of `%s`.